### PR TITLE
[FLINK-18154][Runtime] Check Total Flink Memory plus JVM Metaspace is less than or equal to the configured Total Process Memory

### PR DESCRIPTION
**Pull request by azagrebin**
Check `totalProcessMemorySize >= totalFlinkAndJvmMetaspaceSize` in `ProcessMemoryUtils.deriveJvmMetaspaceAndOverheadFromTotalFlinkMemory` to fail fast and log a descriptive error message.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java [hunk 1]
```diff
@@ -158,6 +156,25 @@ public JvmMetaspaceAndOverhead deriveJvmMetaspaceAndOverheadFromTotalFlinkMemory
 		return jvmMetaspaceAndOverhead;
 	}
 
+	private MemorySize deriveJvmOverheadFromTotalFlinkMemoryAndOtherComponents(
+			Configuration config,
+			MemorySize totalFlinkMemorySize) {
+		MemorySize totalProcessMemorySize = getMemorySizeFromConfig(config, options.getTotalProcessMemoryOption());
+		MemorySize jvmMetaspaceSize = getMemorySizeFromConfig(config, options.getJvmOptions().getJvmMetaspaceOption());
+		MemorySize totalFlinkAndJvmMetaspaceSize = totalFlinkMemorySize.add(jvmMetaspaceSize);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java:164]: It seems as if we are doing these computations twice. Once here and in `deriveJvmMetaspaceAndOverheadFromTotalFlinkMemory` before this method here is called.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java [hunk 2]
```diff
@@ -158,6 +156,25 @@ public JvmMetaspaceAndOverhead deriveJvmMetaspaceAndOverheadFromTotalFlinkMemory
 		return jvmMetaspaceAndOverhead;
 	}
 
+	private MemorySize deriveJvmOverheadFromTotalFlinkMemoryAndOtherComponents(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java:159]: I think the method `deriveProcessSpecWithTotalProcessMemory` could also be susceptible to the same problem.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java [hunk 3]
```diff
@@ -350,4 +368,35 @@ public MemorySize getTotalProcessMemorySize() {
 			throw new UnsupportedOperationException();
 		}
 	}
+
+	private static class DummyFlinkMemoryUtils implements FlinkMemoryUtils<DummyFlinkMemory> {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java:372]: Meta comment: `FlinkMemoryUtils` does not seem to be documented well enough. I would suggest that every interface method should get a JavaDoc explaining what it does. Also the JavaDocs for the implementations of `FlinkMemoryUtils` are scarce at most.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java [hunk 4]
```diff
@@ -350,4 +368,35 @@ public MemorySize getTotalProcessMemorySize() {
...
+		@Override
+		public MemorySize getJvmDirectMemorySize() {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public MemorySize getTotalFlinkMemorySize() {
+			throw new UnsupportedOperationException();
+		}
+	}
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java:401]: I don't really understand why we need these helper classes. Looking at the method bodies it looks as if they are not really needed. The fact that we need to provide them for the test is an indicator that there is something wrong with the code abstraction as the actual change is not easily testable.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java (hunk 2 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java:159]: there is a similar check in the method `deriveJvmMetaspaceAndOverheadWithTotalProcessMemory`, called before:
`jvmMetaspaceAndOverhead.getTotalJvmMetaspaceAndOverheadSize().getBytes() > totalProcessMemorySize.getBytes()`

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java (hunk 4 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java:401]: Indeed, the abstractions are not optimal now for the tests setup.
There is an issue planned for refactoring of memory utils tests: [FLINK-16754](https://issues.apache.org/jira/browse/FLINK-16754). I think we can make a better step towards it in this PR by abstracting only setting of the internal memory. I changed the approach.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java (hunk 3 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java:372]: True, `FlinkMemory` JavaDocs can include its structure and relationships between components. I am not sure whether the methods need any explanation afterwards as they are just getters. The same thing applies to `FlinkMemoryUtils`. Its JavaDocs can be improved. I will add a hot fix commit.

**Comment by azagrebin** [comment]
Thanks for the review @tillrohrmann 
I changed the approach in the test. We should test there one thing under 2 different but standard scenarios: configured fine-grained options and total Flink memory, like in many other memory tests.
I think [FLINK-16754](https://issues.apache.org/jira/browse/FLINK-16754) should abstract this checks better as one-liners.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java [hunk 5]
```diff
@@ -125,6 +125,25 @@ public void testExceptionShouldContainRequiredConfigOptions() {
...
+		conf.set(options.getJvmOptions().getJvmMetaspaceOption(), jvmMetaspaceSize);
+		conf.set(options.getTotalProcessMemoryOption(), totalProcessMemorySize);
+
+		// Total Flink memory + JVM Metaspace > Total Process Memory (no space for JVM overhead)
+
+		configWithFineGrainedOptions(conf, totalFlinkMemorySize);
+		validateFail(conf);
+
+		conf.set(options.getTotalProcessMemoryOption(), totalProcessMemorySize);
+		validateFail(conf);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtilsTestBase.java:144]: I would split this test case into two test cases.

**Pull request closed without merging**